### ISSUE TITLE
Beta Crunchyroll

### ISSUE THREAD

**Issue by federicorosso1993** [ytdl-org/youtube-dl#29290]
## Checklist

- [x] I'm reporting a new site support request
- [x] I've verified that I'm running youtube-dl version **2021.06.06**
- [x] I've checked that all provided URLs are alive and playable in a browser
- [x] I've checked that none of provided URLs violate any copyrights
- [x] I've searched the bugtracker for similar site support requests including closed ones

## Example URLs

https://beta.crunchyroll.com/it/watch/G8WUN2ND8/La-citt-delle-bestie-magiche-coperta-di-neve
https://beta.crunchyroll.com/watch/GJWU2J7Z8/Clash-Class-A-vs-Class-B

## Description

"Beta Crunchyroll" will leave beta on 15 of June (this month).
I saw that another dowloader can support it (but I prefer youtube-dl because it also support many other websites and it can be used from command line), so I think youtube-dl can also support it (at least they don't use netflix/prime drm-like things so it should be possible) and since it will leave beta in two days I think it's time to try to add the support if it's not already done.

**Comment by FoxMcloud5655** [comment]
I can confirm now that the beta is enforced on all logged in users that youtube-dl now does not work when passing in login information via `cookies.txt`.  Free users have not been forced, so if you do not log in, you can still use youtube-dl to download, though only at a very low resolution.

**Comment by SuperSonicHub1** [comment]
Beta Crunchyroll URLs completely compatible with VRV, which youtube-dl **does** have support for. Just replace `beta.crunchyroll.com` with `vrv.co` and you'll be good to go.

**Comment by federicorosso1993** [comment]
> Beta Crunchyroll URLs completely compatible with VRV, which youtube-dl **does** have support for. Just replace `beta.crunchyroll.com` with `vrv.co` and you'll be good to go.

that's not a solution for people outside US since VRV is not like crunchyroll that is also outside US.
For example I cannot use VRV in Italy.
Sure right now the beta website is not mandatory outside US but it will be in the near future.

**Comment by SuperSonicHub1** [comment]
@federicorosso1993 
> that's not a solution for people outside US since VRV is not like crunchyroll that is also outside US.
> For example I cannot use VRV in Italy.

Oh, I forgot about that. I'll look into beta Crunchyroll later today and see what I can do.

**Comment by SuperSonicHub1** [comment]
@FoxMcloud5655
> I can confirm now that the beta is enforced on all logged in users that youtube-dl now does not work when passing in login information via `cookies.txt`. Free users have not been forced, so if you do not log in, you can still use youtube-dl to download, though only at a very low resolution.

I do find it a bit annoying that you can't switch between the two versions of the site once you opt-in but I wasn't aware it affects youtube-dl as well. How irritating.

**Comment by FoxMcloud5655** [comment]
You were able to switch back and forth before June 15th, but it's now enforced.  Such a shame too, as it now looks like a scaled up mobile app, making everything harder to use.  I like their old site better.

**Comment by SuperSonicHub1** [comment]
# How to Crack Crunchyroll Beta (WIP)
## Prerequisites
* Crunchyroll Beta will now be referred to as NuRoll. :^) 
* [VRV extractor](https://github.com/ytdl-org/youtube-dl/blob/master/youtube_dl/extractor/vrv.py): NuRoll's infrastructure is heavily based on Ellation's second project
* [current Crunchyroll extractor](https://github.com/ytdl-org/youtube-dl/blob/master/youtube_dl/extractor/crunchyroll.py): since things like logging in haven't changed yet, we'll be using that code
* You must have a Crunchyroll account to use NuRoll, so all theory crafting will be assuming you're logged in
* NuRoll is seemingly protected with Cloudflare (cookies like `__cf_bm` are present), so we might have to deal with all the issues that will bring
## URL parsing
Just steal [VRV's](https://github.com/ytdl-org/youtube-dl/blob/8fe5d54eb721f1bbb8c8a0d18810a42d1257e406/youtube_dl/extractor/vrv.py#L99). Like I said, infrastructure between the two sites is very similar, to the point that paths are fully compatible (change the domain from `beta.crunchyroll.com` to `vrv.co` and see what happens), so just refer to the VRV extractor.
## Getting Login Tokens
After [logging in with normal Crunchyroll](https://github.com/ytdl-org/youtube-dl/blob/8fe5d54eb721f1bbb8c8a0d18810a42d1257e406/youtube_dl/extractor/crunchyroll.py#L55), a cookie named `etp_rt_cookie` will be set; its value is a UUID. NuRoll will then send a form request in order to get an access token, a la [VRV](https://github.com/ytdl-org/youtube-dl/blob/8fe5d54eb721f1bbb8c8a0d18810a42d1257e406/youtube_dl/extractor/vrv.py#L128) using `etp_rt_cookie`. The dance goes a little like this:
```
POST https://beta-api.crunchyroll.com/auth/v1/token
content-type: application/x-www-form-urlencoded
Authorization: Basic <some token?>
Cookie: etp_rt=<value goes here>;

grant_type=etp_rt_cookie
```
And thusly I have stuck myself. I have no idea where the Basic token comes from nor how it's generated. Looking in the HTML of the NuRoll homepage shows me nothing of importance, so this is where my investigation will end for now. It would be appreciated if someone could pick up the torch and figure out what's happening here.

**Comment by federicorosso1993** [comment]
> It would be appreciated if someone could pick up the torch and figure out what's happening here.

I would love to help but I'm not that good with python or with vb (there is one open source project called crunchyroll downloader v3.0 by hama3254 that was able to implement download from the beta so if someone know vb it should be possible to understand how it works).
I tried to find the line of code that matters based on the post url of the API that you posted on your message, to find how it works and I found this kind of code

```
Public Sub BetaCR(ByVal Auth As String, ByVal Cookie As String)

        Try
            Using client As New WebClient()
                client.Encoding = System.Text.Encoding.UTF8
                client.Headers.Add(My.Resources.ffmpeg_user_agend.Replace(Chr(34), ""))
                client.Headers.Add("ACCEPT: gzip")
                client.Headers.Add("Cookie: " + Cookie)
                client.Headers.Add("Authorization: " + Auth)
                client.Headers.Add("Content-Type: application/x-www-form-urlencoded")
                client.Headers.Add("Referer: https://beta.crunchyroll.com/") '
                Dim reqparm As New Specialized.NameValueCollection

                reqparm.Add("grant_type", "etp_rt_cookie")
                Dim responsebytes = client.UploadValues("https://beta-api.crunchyroll.com/auth/v1/token", "POST", reqparm)
                Dim responsebody = (New Text.UTF8Encoding).GetString(responsebytes)
                'My.Computer.Clipboard.SetText(responsebody)
            End Using
        Catch ex As Exception
            MsgBox(ex.ToString)
        End Try


    End Sub
```
Basically I think is the same Rest call in vb.
Problem is I did not find anything that call that method so I don't really know what kind of value is Auth (so I cannot answer your question). Maybe someone with better understanding of vb can understand where to get that token.
Based on my knowledge of C# and .Net I think that is some kind of value from the User Interface, but I'm not sure, I expected to find a BetaCR function called somewhere but vs code show me nothing.

**Comment by federicorosso1993** [comment]
Now that I think about it that call remind me of an oauth 2.0 post. If that is right the first call should be an url that, once you log in, should return a code and that code should be the Auth bearer code. At least that what I've done with dropbox, one drive and google drive oauth 2.0 API.

**Comment by SuperSonicHub1** [comment]
@federicorosso1993 The login page redirects to NuRoll. The index page for NuRoll, unlike its contemporary VRV, doesn't include a starter bearer. Since I'm very familiar with OAuth due to taking a deep dive with VRV about a month ago, I know that OAuth isn't being used here, it's just a lot of handshaking.

**Comment by meap98** [comment]
do you have an estimation of how long it will take for the downloader to be complete

**Comment by FoxMcloud5655** [comment]
Let's please not bother the devs with such talk.  We hate being asked when things will be done.  Not only do we not know most of the time, but we have to be in the mood and have the time.

This is all volunteer work anyways.  Just like Minecraft mods (where I have most of my experience), it's all done mostly because we want to build something we want to use.  Letting others also use it and learn from it is a bonus.

**Comment by meap98** [comment]
I'm sorry, I'm sorry, I'm sorry, I'm sorry, I'm sorry, I'm sorry, I'm sorry, I'm sorry, I'm sorry, I'm sorry

**Comment by bavellone** [comment]
> And thusly I have stuck myself. I have no idea where the Basic token comes from nor how it's generated. Looking in the HTML of the NuRoll homepage shows me nothing of importance, so this is where my investigation will end for now. It would be appreciated if someone could pick up the torch and figure out what's happening here.

Should be the username and password joined together with a colon and then base64 encoded. See [<credentials>](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Authorization#directives) for more info.

**Comment by SuperSonicHub1** [comment]
@bavellone 
> Should be the username and password joined together with a colon and then base64 encoded. See [credentials](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Authorization#directives) for more info.

Sadly, while the MDN article details how Basic Authentication *should* be used, Crunchyroll doesn't use it that way and therefore `echo "$username:$password" | base64` is not equal to the token used in the browser. Also tried Base64 encoding `etp_rt_cookie` and was still unsuccessful.

**Comment by federicorosso1993** [comment]
I contacted hyugogirubato on discord (@SuperSonicHub1 I reached the same github api page).
He said that he will send a python script that do everything (and use youtube-dl to download).
I'll quote some of his answer to this problem:
"the new api it's a bit complicated lol, you have to connect to your crunchyroll account, then generate a token for each request."
(I wrote this:) SuperSonicHub1 wrote that he cannot find the Authorization: Basic token and that he already tried with $username:$password in base 64.
"yes and no, it's more complicated lol, I will try to make a python script to connect this load the HLS stream links
No, the Basic authorization is simple, it is the Bearer token that is problematic.
Finally do not worry, I will make a program and I will keep you posted.
To use it (connect more easily), I'm not going to use cookies (it's not useful lol), the session_id is enough"

I will keep everyone updated on this.

**Comment by SuperSonicHub1** [comment]
@federicorosso1993 Thanks!

**Comment by hyugogirubato** [comment]
Hello everyone!

As mentioned earlier, @federicorosso1993  contacted me to help you to add support for the site crunchyroll again with the new APIs. Knowing that I master the new API, I will be happy to help you to improve this tool that I appreciate so much! I started to make a program in python "simple" in order to get the necessary to download the videos and subtitles. I’m trying to find as much free time as I can to schedule it, but due to my busy schedule, I think I’ll need another 1-3 days to finish it. I hope this isn’t awkward.

NOTE:the user must be logged in.

**Comment by hyugogirubato** [comment]
Hi!

I just published the python script!

Link: https://github.com/hyugogirubato/Kamyroll-Python

**Comment by SuperSonicHub1** [comment]
@hyugogirubato So the Basic token is baked in? If so, where did you find it within the source of the site?

https://github.com/hyugogirubato/Kamyroll-Python/blob/42b15417dd53b3ca777403f765e16096e1e3fa41/kamyroll.py#L160

**Comment by hyugogirubato** [comment]
Yes, when loading. The Basic token is fixed, it does not change, it is the token to load the pages that changes. @SuperSonicHub1

**Comment by SuperSonicHub1** [comment]
@hyugogirubato Excellent! I'll start working on theory crafting again for youtube-dl ASAP.

**Comment by hyugogirubato** [comment]
Ha great 👍 I can't wait for it to be back!

**Comment by federicorosso1993** [comment]
Awesome!

**Comment by tejing1** [comment]
A note for anyone who's being force to log in (and thus move to the beta UI) due to the age-restriction wall: you don't actually have to log in, just append `?skip_wall=1` to the series url and you're good to go.

**Comment by federicorosso1993** [comment]
I made what I call a "concept conversion" to C#
https://github.com/federicorosso1993/Kamyroll-CSharp

My goal was to understand how the CR beta API works on a programming language that I understand better than python (I can read python but I'm not used to it and for large program like youtube-dl I cannot help a lot) and also to be able to "map" the crunchyroll API in a more "object oriented way" (you don't need to try the call the see what it will return).
It is really basic and I'm not sure I will get more time to do it better (there is the original version for that), it's all in one big file so it's not easy to read (maybe I will fix this), it just take
username and password and the url of the episode you want to download (that API was missing from the python version so I added it).

Not sure if will help anybody but if you need it use it as you please.
I think the original Kamyroll-Python should be all that's needed to implement the support for CR-beta on youtube-dl, but if you don't know the endpoint used with the episode link (since the python version works in a different way) you can find it in my version.

**Comment by hyugogirubato** [comment]
@SuperSonicHub1 I just found something interesting about the new api. Finally, it is not necessary to necessarily be connected (to retrieve the Bearer type access token). To do this, you have to replace etp_rp in the token request by client_id while leaving a session_id in the cookies, and this generates the access token.

Methode: POST
Endpoint: https://beta-api.crunchyroll.com/auth/v1/token
Cookies: {"session_id": SESSION_ID}
Data: {"grant_type": "client_id"}

**Comment by federicorosso1993** [comment]
> @SuperSonicHub1 I just found something interesting about the new api. Finally, it is not necessary to necessarily be connected (to retrieve the Bearer type access token). To do this, you have to replace etp_rp in the token request by client_id while leaving a session_id in the cookies, and this generates the access token.
> 
> Methode: POST
> Endpoint: https://beta-api.crunchyroll.com/auth/v1/token
> Cookies: {"session_id": SESSION_ID}
> Data: {"grant_type": "client_id"}

So the code in your script will not need
 `   headers = {"Authorization": "Basic bm9haWhkZXZtXzZpeWcwYThsMHE6"}`

and will be like

   ```
endpoint = "https://beta-api.crunchyroll.com/auth/v1/token"
    data = {"grant_type": "client_id"}
    cookies = {"session_id": config.get("session_id")}
```

Am I right?

**Comment by SuperSonicHub1** [comment]
> @hyugogirubato 
> Methode: POST
> Endpoint: https://beta-api.crunchyroll.com/auth/v1/token
> Cookies: {"session_id": SESSION_ID}
> Data: {"grant_type": "client_id"}

This is pretty awesome! Thanks for finding this!

**Comment by hyugogirubato** [comment]
@federicorosso1993 & @SuperSonicHub1 

To make the request correctly it is always necessary to use the Basic token.

1. Generate the token without account:

> Methode: POST
> Endpoint: https://beta-api.crunchyroll.com/auth/v1/token
> Headers: {"Authorization": "Basic bm9haWhkZXZtXzZpeWcwYThsMHE6"}
> Cookies = {"session_id": SESSION_ID}
> Data = {"grant_type": "client_id"}

2. Generate token with account:

**Step 1: Retrieve etp_rt from cookies (more reliable than the cookie file but not mandatory)**
> Methode: GET
> Endpoint: https://api.crunchyroll.com/start_session.0.json?session_id=SESSION_ID

**Setp 2: Generate the token with account:**

> Methode: POST
> Endpoint: https://beta-api.crunchyroll.com/auth/v1/token
> Headers: {"Authorization": "Basic bm9haWhkZXZtXzZpeWcwYThsMHE6"}
> Cookies = {"session_id": SESSION_ID, "etp_rt": ETP_RT}
> Data = {"grant_type": "etp_rt_cookie"}

I hope this will help you!

**Comment by federicorosso1993** [comment]
Ok I updated my C# version to be able to download without login. The only thing to add to what hyugogirubato said is that maturity rating should be set to M2, country code should be get from auth/v1/token call and, of course, the channel is always - (since is like the free account).
Everything else stay the same.

**Comment by federicorosso1993** [comment]
hyugogirubato find out that you can get an access_token on this login page https://www.crunchyroll.com/welcome/login or https://www.crunchyroll.com/originals/index.html instead of this https://www.crunchyroll.com/login,
On the welcome/login and originals there is a new API call GET
`https://api.crunchyroll.com/get_upsell_data.0.json?sess_id=1&device_id=whatvalueshouldbeforweb&device_type=com.crunchyroll.static&access_token=giKq5eY27ny3cqz&referer=https:%2F%2Fwww.crunchyroll.com%2Fit%2Fwelcome%2Flogin`

that works with a static device_id whatvalueshouldbeforweb and static access token giKq5eY27ny3cqz
that way it will be really easy to update the access token if it change in the future.
I updated my C# version with a comment for this alternative call (to keep track and docs the new beta API).

Both access_token are good alternative, but I think it is not easy to decompile the desktop/android app to get the access_token if it change (I tried it) while you just need a browser console to get the access_token for the web api call get_upsell_data.
I'm not sure if there is any way to get the same access_token from the regular login page (not really sure why that page does not record the same api call). Of course everything can change on the final version of the beta crunchyroll (but I don't really think they will rewrite the API from scratch.
Maybe youtube-dl can use this call instead of the desktop api call
@SuperSonicHub1 what do you think?

**Comment by SuperSonicHub1** [comment]
We probably shouldn't rely on a mobile API until it's heavily documented. I
imagine most yt-dl contributors don't want want to jump through the hoops
of setting up a proxy in case the API changes.

**Comment by hyugogirubato** [comment]
@SuperSonicHub1 

Following a scrapping of the mobile app, I redid my kamyroll-python program which now uses 100% of the new crunchyroll API. I invite you to watch to tell me if this method to download the videos suits you. Also if the documentation provided is not clear enough, I will be honored if you give me your feedback so that I can improve.

Link: https://github.com/hyugogirubato/Kamyroll-Python

**Comment by federicorosso1993** [comment]
> > @SuperSonicHub1
> > Following a scrapping of the mobile app, I redid my kamyroll-python program which now uses 100% of the new crunchyroll API. I invite you to watch to tell me if this method to download the videos suits you. Also if the documentation provided is not clear enough, I will be honored if you give me your feedback so that I can improve.
> > Link: https://github.com/hyugogirubato/Kamyroll-Python
> 
> ive been reading that ytdl (this project) is "dead" (the owners of the repo stopped merging PRs). this might be a good fit for yt-dlp so you might try submitting it there

you mean this repo?
https://github.com/yt-dlp/yt-dlp

If yes I'll make a site request there too once I verified that it's not working there too (I still need to try it) and I'll link this topic.
I was not aware of the fact that youtube-dl was dead and it was replaced by yt-dlp but I tried to find some info online and that seem to be true (also there was no commit for like 2 months now.

I'm sure that with the code that hyugogirubato made it will really easy to implement beta crunchyroll support into the new youtube-dl.
Thank you.

**Comment by federicorosso1993** [comment]
I made a site request there too
https://github.com/yt-dlp/yt-dlp/issues/915

take a look.

**Comment by tejing1** [comment]
It's worth noting here that I implemented full CR Beta API support (Well, almost full. Movies still don't work. I should get on that.) in yt-dlp a while ago. Their [crunchyroll.py](https://github.com/yt-dlp/yt-dlp/blob/master/yt_dlp/extractor/crunchyroll.py) may be useful for anyone interested in implementing it in youtube-dl. Or you may just want to use yt-dlp instead. Up to you.

**Comment by federicorosso1993** [comment]
> It's worth noting here that I implemented full CR Beta API support (Well, almost full. Movies still don't work. I should get on that.) in yt-dlp a while ago. Their [crunchyroll.py](https://github.com/yt-dlp/yt-dlp/blob/master/yt_dlp/extractor/crunchyroll.py) may be useful for anyone interested in implementing it in youtube-dl. Or you may just want to use yt-dlp instead. Up to you.

Yes I know. I made the request after I made an implementation on my own based on kamyroll
